File: implementations/cpp/minimum_spanning_trees.cpp
/**
 * Reference Implementation: Minimum Spanning Trees (MST)
 * Demonstrates:
 * 1. Kruskal's Algorithm (Global Edge Sorting + Disjoint Set Union).
 * 2. Prim's Algorithm (Frontier Expansion with Min-Priority Queue).
 * 3. Borůvka's Algorithm (Concurrent Component Contraction).
 * 4. Maximum Spanning Tree & Minimum Bottleneck Spanning Tree verification.
 *
 * Language: C++17
 */

#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <queue>
#include <limits>
#include <utility>
#include <cassert>

// ============================================================================
// Data Structures
// ============================================================================
struct Edge {
    int u;
    int v;
    long long weight;
};

struct AdjEdge {
    int to;
    long long weight;
};

struct MstResult {
    long long total_weight;
    std::vector<Edge> edges;
    bool is_connected;
};

struct DSU {
    std::vector<int> parent;
    std::vector<int> rank;

    explicit DSU(int n) : parent(n), rank(n, 0) {
        std::iota(parent.begin(), parent.end(), 0);
    }

    int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    bool unite(int a, int b) {
        a = find(a);
        b = find(b);
        if (a == b) return false;
        if (rank[a] < rank[b]) std::swap(a, b);
        parent[b] = a;
        if (rank[a] == rank[b]) ++rank[a];
        return true;
    }
};

// ============================================================================
// 1. Kruskal's Algorithm
// ============================================================================
MstResult kruskal_mst(int n, std::vector<Edge> edges) {
    if (n <= 1) {
        return {0, {}, true};
    }

    std::sort(edges.begin(), edges.end(), [](const Edge& a, const Edge& b) {
        return a.weight < b.weight;
    });

    DSU dsu(n);
    long long total_weight = 0;
    std::vector<Edge> mst;
    mst.reserve(n - 1);

    for (const auto& e : edges) {
        if (dsu.unite(e.u, e.v)) {
            total_weight += e.weight;
            mst.push_back(e);
            if (static_cast<int>(mst.size()) == n - 1) {
                break;
            }
        }
    }

    bool connected = (static_cast<int>(mst.size()) == n - 1);
    return {total_weight, mst, connected};
}

// ============================================================================
// 2. Prim's Algorithm
// ============================================================================
MstResult prim_mst(int n, const std::vector<std::vector<AdjEdge>>& graph, int start = 0) {
    if (n <= 1) {
        return {0, {}, true};
    }

    std::vector<int> visited(n, 0);
    std::vector<long long> best(n, std::numeric_limits<long long>::max());
    std::vector<int> parent(n, -1);

    using State = std::pair<long long, int>; // {edge_weight, vertex}
    std::priority_queue<State, std::vector<State>, std::greater<State>> pq;

    best[start] = 0;
    pq.push({0, start});

    long long total_weight = 0;
    std::vector<Edge> mst;
    mst.reserve(n - 1);

    while (!pq.empty()) {
        auto [w, u] = pq.top();
        pq.pop();

        if (visited[u]) {
            continue; // Stale heap entry
        }
        visited[u] = 1;

        if (parent[u] != -1) {
            total_weight += w;
            mst.push_back({parent[u], u, w});
        }

        for (const auto& e : graph[u]) {
            int v = e.to;
            if (!visited[v] && e.weight < best[v]) {
                best[v] = e.weight;
                parent[v] = u;
                pq.push({best[v], v});
            }
        }
    }

    bool connected = (static_cast<int>(mst.size()) == n - 1);
    return {total_weight, mst, connected};
}

// ============================================================================
// 3. Borůvka's Algorithm
// ============================================================================
MstResult boruvka_mst(int n, const std::vector<Edge>& edges) {
    if (n <= 1) {
        return {0, {}, true};
    }

    DSU dsu(n);
    int components = n;
    long long total_weight = 0;
    std::vector<Edge> mst;
    mst.reserve(n - 1);

    while (components > 1) {
        std::vector<int> best_edge(n, -1);

        for (int i = 0; i < static_cast<int>(edges.size()); ++i) {
            int u = dsu.find(edges[i].u);
            int v = dsu.find(edges[i].v);
            if (u == v) continue;

            if (best_edge[u] == -1 || edges[i].weight < edges[best_edge[u]].weight) {
                best_edge[u] = i;
            }
            if (best_edge[v] == -1 || edges[i].weight < edges[best_edge[v]].weight) {
                best_edge[v] = i;
            }
        }

        bool merged_any = false;

        for (int i = 0; i < n; ++i) {
            int ei = best_edge[i];
            if (ei == -1) continue;

            const auto& e = edges[ei];
            if (dsu.unite(e.u, e.v)) {
                total_weight += e.weight;
                mst.push_back(e);
                --components;
                merged_any = true;
            }
        }

        if (!merged_any) {
            break;
        }
    }

    bool connected = (static_cast<int>(mst.size()) == n - 1);
    return {total_weight, mst, connected};
}

// ============================================================================
// 4. Maximum Spanning Tree
// ============================================================================
MstResult maximum_spanning_tree(int n, std::vector<Edge> edges) {
    for (auto& e : edges) {
        e.weight = -e.weight;
    }
    MstResult res = kruskal_mst(n, edges);
    res.total_weight = -res.total_weight;
    for (auto& e : res.edges) {
        e.weight = -e.weight;
    }
    return res;
}

// ============================================================================
// Unit Tests & Invariant Verification
// ============================================================================
int main() {
    // ------------------------------------------------------------------------
    // Test 1: Standard 4-Vertex Graph (Arthur's Example)
    // A(0), B(1), C(2), D(3)
    // AB=1, BC=2, CD=3, AC=4, BD=5
    // ------------------------------------------------------------------------
    {
        int n = 4;
        std::vector<Edge> edges = {
            {0, 1, 1},
            {1, 2, 2},
            {2, 3, 3},
            {0, 2, 4},
            {1, 3, 5}
        };

        // Adjacency representation for Prim
        std::vector<std::vector<AdjEdge>> adj(n);
        for (const auto& e : edges) {
            adj[e.u].push_back({e.v, e.weight});
            adj[e.v].push_back({e.u, e.weight});
        }

        MstResult res_k = kruskal_mst(n, edges);
        MstResult res_p = prim_mst(n, adj, 0);
        MstResult res_b = boruvka_mst(n, edges);

        assert(res_k.is_connected);
        assert(res_p.is_connected);
        assert(res_b.is_connected);

        assert(res_k.total_weight == 6);
        assert(res_p.total_weight == 6);
        assert(res_b.total_weight == 6);

        assert(res_k.edges.size() == 3);
        assert(res_p.edges.size() == 3);
        assert(res_b.edges.size() == 3);

        // Maximum Spanning Tree: edges with weights 5, 4, 3 => weight = 12
        MstResult res_max = maximum_spanning_tree(n, edges);
        assert(res_max.is_connected);
        assert(res_max.total_weight == 12);
    }

    // ------------------------------------------------------------------------
    // Test 2: Disconnected Graph
    // ------------------------------------------------------------------------
    {
        int n = 5;
        // Component 1: {0, 1}, Component 2: {2, 3, 4}
        std::vector<Edge> edges = {
            {0, 1, 10},
            {2, 3, 5},
            {3, 4, 7},
            {2, 4, 8}
        };

        std::vector<std::vector<AdjEdge>> adj(n);
        for (const auto& e : edges) {
            adj[e.u].push_back({e.v, e.weight});
            adj[e.v].push_back({e.u, e.weight});
        }

        MstResult res_k = kruskal_mst(n, edges);
        MstResult res_p = prim_mst(n, adj, 0);
        MstResult res_b = boruvka_mst(n, edges);

        assert(!res_k.is_connected);
        assert(!res_p.is_connected);
        assert(!res_b.is_connected);
    }

    // ------------------------------------------------------------------------
    // Test 3: Trivial Single-Vertex Graph
    // ------------------------------------------------------------------------
    {
        int n = 1;
        std::vector<Edge> edges = {};
        std::vector<std::vector<AdjEdge>> adj(1);

        MstResult res_k = kruskal_mst(n, edges);
        MstResult res_p = prim_mst(n, adj, 0);
        MstResult res_b = boruvka_mst(n, edges);

        assert(res_k.is_connected && res_k.total_weight == 0 && res_k.edges.empty());
        assert(res_p.is_connected && res_p.total_weight == 0 && res_p.edges.empty());
        assert(res_b.is_connected && res_b.total_weight == 0 && res_b.edges.empty());
    }

    // ------------------------------------------------------------------------
    // Test 4: Complete Graph K4 with Non-Distinct Weights
    // ------------------------------------------------------------------------
    {
        int n = 4;
        std::vector<Edge> edges = {
            {0, 1, 2}, {0, 2, 2}, {0, 3, 3},
            {1, 2, 2}, {1, 3, 4}, {2, 3, 2}
        };
        std::vector<std::vector<AdjEdge>> adj(n);
        for (const auto& e : edges) {
            adj[e.u].push_back({e.v, e.weight});
            adj[e.v].push_back({e.u, e.weight});
        }

        MstResult res_k = kruskal_mst(n, edges);
        MstResult res_p = prim_mst(n, adj, 0);
        MstResult res_b = boruvka_mst(n, edges);

        assert(res_k.is_connected && res_p.is_connected && res_b.is_connected);
        assert(res_k.total_weight == 6); // 3 edges of weight 2
        assert(res_p.total_weight == 6);
        assert(res_b.total_weight == 6);
    }

    std::cout << "[PASS] All Minimum Spanning Tree C++ unit tests passed.\n";
    return 0;
}

File: implementations/python/minimum_spanning_trees.py
"""
Reference Implementation: Minimum Spanning Trees (MST)
Demonstrates:
1. Kruskal's Algorithm (Global Edge Sorting + Disjoint Set Union).
2. Prim's Algorithm (Frontier Expansion with Min-Priority Queue).
3. Borůvka's Algorithm (Concurrent Component Contraction).
4. Maximum Spanning Tree & Minimum Bottleneck Spanning Tree verification.

Language: Python 3
"""

import heapq
from typing import List, Optional, Tuple
import unittest


class DSU:
    """Disjoint Set Union with path compression and union by rank."""

    def __init__(self, n: int):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x: int) -> int:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def unite(self, a: int, b: int) -> bool:
        a = self.find(a)
        b = self.find(b)
        if a == b:
            return False
        if self.rank[a] < self.rank[b]:
            a, b = b, a
        self.parent[b] = a
        if self.rank[a] == self.rank[b]:
            self.rank[a] += 1
        return True


def kruskal_mst(
    n: int,
    edges: List[Tuple[int, int, int]],
) -> Tuple[Optional[int], List[Tuple[int, int, int]]]:
    """Computes the MST using Kruskal's algorithm.

    Time Complexity: O(E log E)
    Space Complexity: O(V)
    """
    if n <= 1:
        return 0, []

    sorted_edges = sorted(edges, key=lambda e: e[2])
    dsu = DSU(n)
    total_weight = 0
    mst = []

    for u, v, w in sorted_edges:
        if dsu.unite(u, v):
            total_weight += w
            mst.append((u, v, w))
            if len(mst) == n - 1:
                break

    if len(mst) != n - 1:
        return None, []

    return total_weight, mst


def prim_mst(
    graph: List[List[Tuple[int, int]]],
    start: int = 0,
) -> Tuple[Optional[int], List[Tuple[int, int, int]]]:
    """Computes the MST using Prim's algorithm with a min-priority queue.

    Time Complexity: O((V + E) log V)
    Space Complexity: O(V + E)
    """
    n = len(graph)
    if n <= 1:
        return 0, []

    visited = [False] * n
    best = [10**18] * n
    parent = [-1] * n

    pq = [(0, start)]
    best[start] = 0

    total_weight = 0
    mst_edges = []

    while pq:
        w, u = heapq.heappop(pq)

        if visited[u]:
            continue
        visited[u] = True

        if parent[u] != -1:
            total_weight += w
            mst_edges.append((parent[u], u, w))

        for v, weight in graph[u]:
            if not visited[v] and weight < best[v]:
                best[v] = weight
                parent[v] = u
                heapq.heappush(pq, (best[v], v))

    if len(mst_edges) != n - 1:
        return None, []

    return total_weight, mst_edges


def boruvka_mst(
    n: int,
    edges: List[Tuple[int, int, int]],
) -> Tuple[Optional[int], List[Tuple[int, int, int]]]:
    """Computes the MST using Borůvka's algorithm.

    Time Complexity: O(E log V)
    Space Complexity: O(V)
    """
    if n <= 1:
        return 0, []

    dsu = DSU(n)
    components = n
    total_weight = 0
    mst = []

    while components > 1:
        best_edge = [-1] * n

        for i, (u, v, w) in enumerate(edges):
            cu = dsu.find(u)
            cv = dsu.find(v)
            if cu == cv:
                continue

            if best_edge[cu] == -1 or w < edges[best_edge[cu]][2]:
                best_edge[cu] = i
            if best_edge[cv] == -1 or w < edges[best_edge[cv]][2]:
                best_edge[cv] = i

        merged_any = False

        for i in range(n):
            ei = best_edge[i]
            if ei == -1:
                continue
            u, v, w = edges[ei]
            if dsu.unite(u, v):
                total_weight += w
                mst.append((u, v, w))
                components -= 1
                merged_any = True

        if not merged_any:
            break

    if len(mst) != n - 1:
        return None, []

    return total_weight, mst


def maximum_spanning_tree(
    n: int,
    edges: List[Tuple[int, int, int]],
) -> Tuple[Optional[int], List[Tuple[int, int, int]]]:
    """Computes the Maximum Spanning Tree by negating edge weights."""
    neg_edges = [(u, v, -w) for u, v, w in edges]
    cost, mst = kruskal_mst(n, neg_edges)
    if cost is None:
        return None, []
    return -cost, [(u, v, -w) for u, v, w in mst]


class TestMinimumSpanningTrees(unittest.TestCase):
    def test_standard_4_vertex_graph(self):
        n = 4
        edges = [
            (0, 1, 1),
            (1, 2, 2),
            (2, 3, 3),
            (0, 2, 4),
            (1, 3, 5),
        ]
        adj = [[] for _ in range(n)]
        for u, v, w in edges:
            adj[u].append((v, w))
            adj[v].append((u, w))

        w_k, mst_k = kruskal_mst(n, edges)
        w_p, mst_p = prim_mst(adj, 0)
        w_b, mst_b = boruvka_mst(n, edges)

        self.assertEqual(w_k, 6)
        self.assertEqual(w_p, 6)
        self.assertEqual(w_b, 6)

        self.assertEqual(len(mst_k), 3)
        self.assertEqual(len(mst_p), 3)
        self.assertEqual(len(mst_b), 3)

        w_max, mst_max = maximum_spanning_tree(n, edges)
        self.assertEqual(w_max, 12)
        self.assertEqual(len(mst_max), 3)

    def test_disconnected_graph(self):
        n = 5
        edges = [
            (0, 1, 10),
            (2, 3, 5),
            (3, 4, 7),
            (2, 4, 8),
        ]
        adj = [[] for _ in range(n)]
        for u, v, w in edges:
            adj[u].append((v, w))
            adj[v].append((u, w))

        w_k, _ = kruskal_mst(n, edges)
        w_p, _ = prim_mst(adj, 0)
        w_b, _ = boruvka_mst(n, edges)

        self.assertIsNone(w_k)
        self.assertIsNone(w_p)
        self.assertIsNone(w_b)

    def test_single_vertex(self):
        n = 1
        edges = []
        adj = [[]]

        w_k, mst_k = kruskal_mst(n, edges)
        w_p, mst_p = prim_mst(adj, 0)
        w_b, mst_b = boruvka_mst(n, edges)

        self.assertEqual(w_k, 0)
        self.assertEqual(w_p, 0)
        self.assertEqual(w_b, 0)
        self.assertEqual(len(mst_k), 0)
        self.assertEqual(len(mst_p), 0)
        self.assertEqual(len(mst_b), 0)

    def test_k4_duplicate_weights(self):
        n = 4
        edges = [
            (0, 1, 2),
            (0, 2, 2),
            (0, 3, 3),
            (1, 2, 2),
            (1, 3, 4),
            (2, 3, 2),
        ]
        adj = [[] for _ in range(n)]
        for u, v, w in edges:
            adj[u].append((v, w))
            adj[v].append((u, w))

        w_k, _ = kruskal_mst(n, edges)
        w_p, _ = prim_mst(adj, 0)
        w_b, _ = boruvka_mst(n, edges)

        self.assertEqual(w_k, 6)
        self.assertEqual(w_p, 6)
        self.assertEqual(w_b, 6)


if __name__ == "__main__":
    unittest.main()
